exerciser/src/main.rs

```diff
@@ -0,0 +1,155 @@
+// limitations under the License.
+
+use anyhow::Context;
+use log::{info, trace};
+use pulldown_cmark::{CowStr, Event, Parser, Tag};
+use std::{
+    env::args,
+    fs::{create_dir, create_dir_all, read_to_string, File},
+    io::Write,
+    path::Path,
+    process::exit,
+};
+
+const INCLUDE_START: &str = "{{#include ";
+const INCLUDE_END: &str = "}}";
+
+fn main() -> anyhow::Result<()> {
+    pretty_env_logger::init();
+
+    let args = args().collect::<Vec<_>>();
+
+    if args.len() != 3 {
+        eprintln!("Usage:");
+        eprintln!(
+            "  {} <src/exercises/exercise.md> <output directory>",
+            args[0]
+        );
+        exit(1);
+    }
+
+    let input_filename = Path::new(&args[1]);
+    let output_directory = Path::new(&args[2]);
+
+    create_dir(output_directory).with_context(|| {
+        format!("Failed to create output directory {:?}", output_directory)
+    })?;
+
+    let input_directory = input_filename
+        .parent()
+        .with_context(|| "Input file has no parent directory.")?;
+    let input_contents = read_to_string(input_filename)
+        .with_context(|| format!("Failed to open {:?}", input_filename))?;
+    let parser = Parser::new(&input_contents);
+
+    let mut next_filename: Option<String> = None;
+    let mut current_file: Option<File> = None;
+    for event in parser {
+        trace!("{:?}", event);
+        match event {
+            Event::Code(x) => {
+                info!("{}:", x);
+                next_filename = Some(x.into_string());
+            }
+            Event::Start(Tag::CodeBlock(x)) => {
+                info!("Start {:?}", x);
+                if let Some(filename) = &next_filename {
+                    let full_filename = output_directory.join(filename);
+                    info!("Opening {:?}", full_filename);
+                    if let Some(directory) = full_filename.parent() {
+                        create_dir_all(directory)?;
+                    }
+                    current_file = Some(File::create(full_filename)?);
+                    next_filename = None;
+                }
+            }
+            Event::Text(text) => {
+                info!("Text: {:?}", text);
+                if let Some(output_file) = &mut current_file {
+                    write_output(text, input_directory, output_file)?;
+                }
+            }
+            Event::End(Tag::CodeBlock(x)) => {
+                info!("End   {:?}", x);
+                current_file = None;
+            }
+            _ => {}
+        }
+    }
+
+    Ok(())
+}
+
+/// Writes the given output file based on the given code text from the Markdown input, processing
+/// include directives as necessary.
+fn write_output(
+    text: CowStr,
+    input_directory: &Path,
+    output_file: &mut File,
+) -> anyhow::Result<()> {
+    for line in text.lines() {
+        info!("Line: {:?}", line);
+        if let (Some(start), Some(end)) =
```

Comment:
Unfortunately it looks like it's not possible to run an mdbook preprocessor by itself without processing the whole book. It would be possible to do that, but it would make the process more complicated, and I'd rather keep this as a fairly simple lightweight tool if possible.

Reply:
> Unfortunately it looks like it's not possible to run an mdbook preprocessor by itself without processing the whole book.

Right, I was simply thinking you would call `mdbook build` with the right option to make it use the `markdown` renderer, then inspect the generated Markdown files in your temporary build output directory.

That would be an external tool which could be called as part of the publishing process -- instead of a tool which is called all the time like the renderer we have now.

Reply:
> Unfortunately it looks like it's not possible to run an mdbook preprocessor by itself without processing the whole book.

Right, I was simply thinking you would call `mdbook build` with the right option to make it use the `markdown` renderer, then inspect the generated Markdown files in your temporary build output directory.

That would be an external tool which could be called as part of the publishing process -- instead of a tool which is called all the time like the renderer we have now.